### ISSUE TITLE
Clarification re: supported data types in `jax.linearize` and `jax.linear_transpose`

### ISSUE THREAD

**Issue by johannahaffner** [jax-ml/jax#25517]
Hi All, 

IIUC `jax.linearize` and `jax.linear_transpose` do not support taking derivatives with respect to integer values, i.e. there is no forward-mode analogy of `jax.grad(..., allow_int=True)`. In JAX forward-mode autodiff, integers are treated as constants. 

When transposing the linearised function taking an integer input, the following error is raised

>ValueError: linearized function called on tangent values inconsistent with the original primal values: got ShapedArray(int32[], weak_type=True) for primal aval ShapedArray(int32[], weak_type=True)

Could we change that error message to something more informative by special-casing integer inputs? I'm happy to do it and open a PR.

```python
import jax

def fn(x):
    return x**2

y_float = 3.0
y_int = 3

fn_eval_float, lin_fn_float = jax.linearize(fn, 3.0)
fn_eval_int, lin_fn_int = jax.linearize(fn, 3)

grad_float = jax.linear_transpose(lin_fn_float, y_float)(1.0)  # Works fine
grad_int = jax.linear_transpose(lin_fn_int, y_int)(1.0)  # Raises value error
```

This issue came up [downstream](https://github.com/patrick-kidger/optimistix/issues/96#issuecomment-2546416892).

**Comment by jakevdp** [comment]
Assigning @dougalm, who has been working on linearization improvements (cf. #25864)

**Comment by dougalm** [comment]
Yes it's a good idea to mention the expected tangent types. Here's a PR that does that: #25875.

With that, the error message now reads:
```
ValueError: linearized function called on tangent values inconsistent with the original primal values: got tange
nt aval ShapedArray(int32[], weak_type=True) for primal aval ShapedArray(int32[], weak_type=True) but expected S
hapedArray(float0[], weak_type=True) 
```
How does that look?
